File: python/actions.py
"""
Side-effect actions that fire after the user decides on a request.

Supported per type config:
  forward_to:      email address — forwards the original email to this address
  webhook:         URL — POSTs the request JSON
  reply_to_sender: string template — reply to the original sender's thread
                   (supports {field_name} interpolation from the extracted fields)
"""

import json
import re
import urllib.error
import urllib.request


def _interpolate(template: str, fields: dict) -> str:
    """Replace {key} with fields[key], unbraced for missing keys."""
    def replace(m):
        key = m.group(1)
        return str(fields.get(key, "")) or "(unspecified)"
    return re.sub(r"\{(\w+)\}", replace, template)


def _post_webhook(url: str, payload: dict) -> bool:
    try:
        req = urllib.request.Request(
            url,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=10):
            pass
        return True
    except Exception as e:
        print(f"  ! webhook failed: {e}")
        return False


def fire(*, agentmail_client, inbox, request_row: dict, type_config,
         decision: str, decision_args: dict | None = None) -> list[str]:
    """Run the configured side-effect actions for the decision.
    Returns a list of human-readable confirmations to include in the user ack."""
    actions_block = type_config.approve if decision == "approve" else type_config.decline
    fields = json.loads(request_row.get("fields_json", "{}") or "{}")
    confirmations: list[str] = []

    # 1. forward_to
    forward_to = actions_block.get("forward_to")
    if forward_to:
        try:
            agentmail_client.inboxes.messages.forward(
                inbox.inbox_id, request_row["source_message_id"],
                to=[forward_to],
                text=(
                    f"[Approval inbox - {decision.upper()}]\n\n"
                    f"Type: {request_row['type']}\n"
                    f"Summary: {request_row['summary']}\n"
                    f"Decided by: {decision}\n\n"
                    f"Fields: {json.dumps(fields, indent=2)}\n\n"
                    f"Original email forwarded below."
                ),
            )
            confirmations.append(f"forwarded to {forward_to}")
        except Exception as e:
            print(f"  ! forward failed: {e}")
            confirmations.append(f"forward to {forward_to} FAILED ({e})")

    # 2. webhook
    webhook = actions_block.get("webhook")
    if webhook:
        ok = _post_webhook(webhook, {
            "request_id": request_row["id"],
            "type": request_row["type"],
            "decision": decision,
            "summary": request_row["summary"],
            "fields": fields,
            "source_sender": request_row.get("source_sender", ""),
            "decided_text": (decision_args or {}).get("text", ""),
        })
        confirmations.append(f"webhook {'fired' if ok else 'FAILED'}")

    # 3. reply_to_sender
    reply_template = actions_block.get("reply_to_sender")
    if reply_template:
        sender = request_row.get("source_sender", "")
        if not sender:
            confirmations.append("reply_to_sender skipped (no source sender on record)")
        else:
            try:
                # Interpolate fields into the template
                merged = {**fields,
                          "decision": decision,
                          "reason": (decision_args or {}).get("reason", ""),
                          "changes_text": (decision_args or {}).get("changes_text", "")}
                body = _interpolate(reply_template, merged)
                agentmail_client.inboxes.messages.reply(
                    inbox.inbox_id, request_row["source_message_id"],
                    text=body,
                )
                confirmations.append(f"replied to {sender}")
            except Exception as e:
                print(f"  ! reply_to_sender failed: {e}")
                confirmations.append(f"reply to {sender} FAILED ({e})")

    if not confirmations:
        confirmations.append("(no side-effect actions configured for this decision)")

    return confirmations

File: agentmail-approval-inbox/typescript/src/actions.ts
/**
 * Side-effect actions: forward_to / webhook / reply_to_sender.
 */

import { TypeConfig, ActionBlock } from "./typesConfig.js";
import { RequestRow } from "./requestsStore.js";

function interpolate(template: string, fields: Record<string, any>): string {
  return template.replace(/\{(\w+)\}/g, (_, key) => {
    const v = fields[key];
    return (v ?? "") === "" ? "(unspecified)" : String(v);
  });
}

async function postWebhook(url: string, payload: Record<string, any>): Promise<boolean> {
  try {
    await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    return true;
  } catch (e: any) {
    console.warn(`  ! webhook failed: ${e.message}`);
    return false;
  }
}

export async function fire(opts: {
  agentmail: any;
  inbox: any;
  requestRow: RequestRow;
  typeConfig: TypeConfig;
  decision: "approve" | "decline";
  decisionArgs?: { reason?: string; changes_text?: string };
}): Promise<string[]> {
  const block: ActionBlock = opts.decision === "approve" ? opts.typeConfig.approve : opts.typeConfig.decline;
  let fields: Record<string, any> = {};
  try { fields = JSON.parse(opts.requestRow.fields_json || "{}"); } catch {}
  const confirmations: string[] = [];

  // 1. forward_to
  if (block.forward_to) {
    try {
      await opts.agentmail.inboxes.messages.forward(opts.inbox.inboxId, opts.requestRow.source_message_id, {
        to: [block.forward_to],
        text:
          `[Approval inbox - ${opts.decision.toUpperCase()}]\n\n` +
          `Type: ${opts.requestRow.type}\n` +
          `Summary: ${opts.requestRow.summary}\n` +
          `Decided by: ${opts.decision}\n\n` +
          `Fields: ${JSON.stringify(fields, null, 2)}\n\n` +
          `Original email forwarded below.`,
      });
      confirmations.push(`forwarded to ${block.forward_to}`);
    } catch (e: any) {
      console.warn(`  ! forward failed: ${e.message}`);
      confirmations.push(`forward to ${block.forward_to} FAILED (${e.message})`);
    }
  }

  // 2. webhook
  if (block.webhook) {
    const ok = await postWebhook(block.webhook, {
      request_id: opts.requestRow.id,
      type: opts.requestRow.type,
      decision: opts.decision,
      summary: opts.requestRow.summary,
      fields,
      source_sender: opts.requestRow.source_sender,
      decided_text: opts.decisionArgs?.reason ?? opts.decisionArgs?.changes_text ?? "",
    });
    confirmations.push(`webhook ${ok ? "fired" : "FAILED"}`);
  }

  // 3. reply_to_sender
  if (block.reply_to_sender) {
    const sender = opts.requestRow.source_sender;
    if (!sender) {
      confirmations.push("reply_to_sender skipped (no source sender on record)");
    } else {
      try {
        const merged = {
          ...fields,
          decision: opts.decision,
          reason: opts.decisionArgs?.reason ?? "",
          changes_text: opts.decisionArgs?.changes_text ?? "",
        };
        const body = interpolate(block.reply_to_sender, merged);
        await opts.agentmail.inboxes.messages.reply(opts.inbox.inboxId, opts.requestRow.source_message_id, {
          text: body,
        });
        confirmations.push(`replied to ${sender}`);
      } catch (e: any) {
        console.warn(`  ! reply_to_sender failed: ${e.message}`);
        confirmations.push(`reply to ${sender} FAILED (${e.message})`);
      }
    }
  }

  if (!confirmations.length) {
    confirmations.push("(no side-effect actions configured for this decision)");
  }
  return confirmations;
}
